### PR TITLE
Created ProfileInformation Service

### PR DESCRIPTION
**Pull request by RachalCassity**
**Note**: Delete the description statements, complete each step. **None are optional**, but can be justified as to why they cannot be completed as written. Provide known gaps to testing that may raise the risk of merging to production.

## Summary

Created new ProfileInformation::Service to replace ContactInformation::Service.

Changes:
VAProfileRedis::ProfileInformation is replacing VAProfileRedis::ContactInformation. 

The model method, transaction_response_class, will be removed in the next pull request regarding transactions.

## Profile Information Issues
Questions for Auth Exper:
1) Please check vcr paths: 
        methods:` :put or :post,`  
         uri: `https://int.vet360.va.gov/profile-service/profile/v2/addresses`

## Related issue(s)

- department-of-veterans-affairs/va.gov-team#87645

## Testing done

- [x] *New code is covered by unit tests*

## Screenshots
_Note: Optional_

## What areas of the site does it impact?
* None, flipper is not on

## Acceptance criteria

- [x]  I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [x]  No error nor warning in the console.
- [x]  Events are being sent to the appropriate logging solution
- [x]  Documentation has been updated (link to documentation)
- [x]  No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [x]  Feature/bug has a monitor built into Datadog (if applicable)
- [x]  If app impacted requires authentication, did you login to a local build and verify all authenticated routes work as expected
- [x]  I added a screenshot of the developed feature

## Requested Feedback

(OPTIONAL)_What should the reviewers know in addition to the above. Is there anything specific you wish the reviewer to assist with. Do you have any concerns with this PR, why?_

### CODE REVIEW

**Comment by stevenjcumming** [comment]
@RachalCassity a bad test was introduced into master. Could you update your branch to resolve the failing code check?

File: lib/va_profile/models/associated_person.rb [hunk 1]
```diff
@@ -38,6 +38,14 @@ class AssociatedPerson < Base
       attribute :state, String
       attribute :zip_code, String
       attribute :primary_phone, String
+
+      # def self.bio_path
```
Comment (LindseySaari) [inline lib/va_profile/models/associated_person.rb:42]: Does this need to be uncommented?

File: lib/va_profile/profile_information/service.rb [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+      # Determine if the record needs to be created or updated with reassign_http_verb
+      # Ensure http_verb is a symbol for the response request
+      def create_or_update_info(http_verb, record)
+        with_monitoring do
+          icn_with_aaid_present!
+          model = record.class
+          http_verb = http_verb.to_sym == :update ? reassign_http_verb(record) : http_verb.to_sym
+          update_path = record.try(:permission_type).present? ? 'permissions' : record.contact_info_attr.pluralize
+          raw_response = perform(http_verb, update_path, record.in_json)
+          response = model.transaction_response_class.from(raw_response)
```
Comment (LindseySaari) [inline lib/va_profile/profile_information/service.rb:106]: Given that this implementation relies on duck typing and the presence of specific methods on the models, etc, Could you explicitly call out these methods as required in the base class (via comments or something). This ensures that all subclasses implement these necessary methods for the functionality to work correctly. The base class itself doesn't enforce how these methods are implemented, only that they must exist.

File: lib/va_profile/models/telephone.rb [hunk 3]
```diff
@@ -146,6 +146,32 @@ def self.build_from(body)
...
+
+      def self.send_change_notifcations?
+        true
+      end
+
+      def contact_info_attr
+        'telephone'
+      end
+
+      def self.contact_info_attr(record)
```
Comment (LindseySaari) [inline lib/va_profile/models/telephone.rb:166]: This might look a little cleaner as a CASE statement

File: lib/va_profile/profile_information/service.rb [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+        residence_address: 'Home address',
+        correspondence_address: 'Mailing address',
+        email: 'Email address',
+        phone: 'Phone number',
+        home_phone: 'Home phone number',
+        mobile_phone: 'Mobile phone number',
+        work_phone: 'Work phone number'
+      }.freeze
+
+      # def get_response(type)
```
Comment (LindseySaari) [inline lib/va_profile/profile_information/service.rb:34]: Can this be removed as well? We always have git history if it needs to be resurrected

File: lib/va_profile/profile_information/service.rb [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+
+        record.id = contact_info.public_send(attr)&.id
+        record.id.present? ? :put : :post
+      end
+
+      def get_email_personalisation(type)
+        { 'contact_info' => EMAIL_PERSONALISATIONS[type] }
+      end
+
+      # def send_change_notifications(transaction_status)
```
Comment (LindseySaari) [inline lib/va_profile/profile_information/service.rb:176]: Is there a reason the following methods are commented out? Looks like it hooks into VANotify to send notifications

File: lib/va_profile/profile_information/transaction_response.rb [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+
+      def self.log_mpi_error
+        if error?
+          @response_body['tx_messages'].each do |tx_message|
+            if tx_message['code'] == NOT_FOUND_IN_MPI_CODE
+              return log_message_to_sentry(
+                'va profile mpi not found',
+                :error,
+                {
+                  icn: @user.icn,
```
Comment (LindseySaari) [inline lib/va_profile/profile_information/transaction_response.rb:115]: edipi and icn are considered PII... are these masked when logged to Sentry?

File: lib/va_profile/profile_information/service.rb [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+        elsif e.status >= 400 && e.status < 500
+          return PersonResponse.new(e.status, person: nil)
+        end
+
+        handle_error(e)
+      rescue => e
+        handle_error(e)
+      end
+
+      def self.get_person(vet360_id)
```
Comment (LindseySaari) [inline lib/va_profile/profile_information/service.rb:87]: Does the initialize method for this class's constructor accepts an object with arbitrary attributes (the stub user struct)? It seems like [it expects a user on the base class](https://github.com/department-of-veterans-affairs/vets-api/blob/master/lib/va_profile/service.rb#L15-L17) unless I'm missing something

File: lib/va_profile/profile_information/transaction_response.rb [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+
+      attr_reader :response_body
+
+      def self.from(raw_response = nil, _user = nil)
+        @response_body = raw_response&.body
+        if error?
+          log_message_to_sentry(
+            'VAProfile transaction error',
+            :error,
+            { response_body: @response_body },
```
Comment (LindseySaari) [inline lib/va_profile/profile_information/transaction_response.rb:22]: Will response_body have any PII when logging to Sentry?

File: lib/va_profile/profile_information/transaction_response.rb [hunk 9]
```diff
@@ -0,0 +1,194 @@
+# frozen_string_literal: true
+
+require 'va_profile/models/transaction'
+require 'va_profile/response'
+
+module VAProfile
```
Comment (LindseySaari) [inline lib/va_profile/profile_information/transaction_response.rb:6]: The classes defined here also use a duck typing pattern, so maybe we want to add some comments around the required duck typing methods that are assumed and being used elsewhere

File: config/settings.yml [hunk 10]
```diff
@@ -273,6 +273,11 @@ vet360:
     api_key: "<AV_KEY>"
   health_benefit:
     mock: true
+  profile_information:
+    cache_enabled: true
```
Comment (LindseySaari) [inline config/settings.yml:277]: Where is cache_enabled used?

File: lib/va_profile/models/associated_person.rb (hunk 1 above)
Comment (RachalCassity) [reply lib/va_profile/models/associated_person.rb:42]: Not yet. This will be uncommented after the upgarde

File: lib/va_profile/profile_information/service.rb (hunk 7 above)
Comment (RachalCassity) [reply lib/va_profile/profile_information/service.rb:87]: This will be removed in when merging the other services.

File: config/settings.yml (hunk 10 above)
Comment (RachalCassity) [reply config/settings.yml:277]: When cacheing redis. You can see it throughout,  VAProfile::Configuration::SETTINGS.profile_infromation.cache_enabled

Comment (RachalCassity) [reply config/settings.yml:277]: The setting will be need to be reformatted in the future

File: lib/va_profile/profile_information/service.rb (hunk 5 above)
Comment (RachalCassity) [reply lib/va_profile/profile_information/service.rb:176]: These will be uncommented in other PRs :)

File: lib/va_profile/profile_information/transaction_response.rb (hunk 8 above)
Comment (RachalCassity) [reply lib/va_profile/profile_information/transaction_response.rb:22]: I am not sure. I am assuming not since it's already being sent

File: lib/va_profile/profile_information/transaction_response.rb (hunk 6 above)
Comment (RachalCassity) [reply lib/va_profile/profile_information/transaction_response.rb:115]: Will look into this.

File: lib/va_profile/profile_information/service.rb (hunk 2 above)
Comment (RachalCassity) [reply lib/va_profile/profile_information/service.rb:106]: This is due to the create_or_update action being called to vets-api. In this method the create_or_update is reassigned in the reassign_http_verb method. the reassign_http_verb looks to see if the record exists or does not exists.

File: .rubocop.yml [hunk 11]
```diff
@@ -77,7 +77,7 @@ Style/ZeroLengthPredicate:
   Enabled: false
 
 Layout/LineLength:
-  Max: 120
+  Max: 125
```
Comment (bosawt) [inline .rubocop.yml:80]: Why was the line length updated?

Comment (RachalCassity) [reply .rubocop.yml:80]: It was reverted.

Comment (bosawt) [reply .rubocop.yml:80]: What do you mean?

File: app/controllers/concerns/vet360/writeable.rb [hunk 12]
```diff
@@ -2,6 +2,7 @@
 
 require 'common/exceptions/validation_errors'
 require 'va_profile/contact_information/service'
+require 'va_profile/profile_information/service'
 
 module Vet360
```
Comment (bosawt) [inline app/controllers/concerns/vet360/writeable.rb:7]: Something to consider. I know this is working off of existing code and the intention might be to update as little as possible while moving things over to the new service, but I wonder if things would be simpler if this entire concern was removed and the controllers just call the models directly? The complexity was already there before this PR but it seems like this concern just complicates things without DRYing up much at all

File: .rubocop.yml (hunk 11 above)
Comment (RachalCassity) [reply .rubocop.yml:84]: I will put this PR back in draft for now.

File: app/controllers/concerns/vet360/writeable.rb (hunk 12 above)
Comment (RachalCassity) [reply app/controllers/concerns/vet360/writeable.rb:7]: Yes, I agree. I will put this in the notes after this upgrade and we start to combine VAProfile services. This will no longer be needed.

File: lib/va_profile/models/telephone.rb (hunk 3 above)
Comment (RachalCassity) [reply lib/va_profile/models/telephone.rb:166]: Fixing now.

File: app/controllers/concerns/vet360/writeable.rb (hunk 12 above)
Comment (RachalCassity) [reply app/controllers/concerns/vet360/writeable.rb:7]: Same with transactionable.rb

File: lib/va_profile/profile_information/service.rb (hunk 4 above)
Comment (RachalCassity) [reply lib/va_profile/profile_information/service.rb:34]: Yeah, i can remove it for now :)

File: lib/va_profile/models/address.rb [hunk 13]
```diff
@@ -97,6 +97,28 @@ def self.build_from(body)
...
+
+      def self.transaction_response_class
+        VAProfile::ProfileInformation::AddressTransactionResponse
+      end
+
+      def self.transaction_status_path(user, transaction_id)
+        "#{user.vet360_id}/addresses/status/#{transaction_id}"
+      end
+
+      def self.send_change_notifcations?
```
[the remark this answers is not in the document]
Comment (rmtolmach) [reply lib/va_profile/models/address.rb:109]: Trevor is calling out the missing `i` in "notifcations"

File: .rubocop.yml [hunk 14]
```diff
@@ -64,6 +64,9 @@ RSpec/VariableName:
 Style/GuardClause:
   Enabled: false
 
+Style/IfUnlessModifier:
```
Comment (kpethtel) [inline .rubocop.yml:67]: Why are we changing this rule? It seems unrelated to the work.

**Review by kpethtel — COMMENTED** [review]
My review was requested, but I don't know enough about this to review. I am curious why we would change the linting rule, though.

File: lib/va_profile/models/address.rb (hunk 13 above)
[the remark this answers is not in the document]
Comment (dcloud) [reply lib/va_profile/models/address.rb:109]:
```suggestion
      def self.send_change_notifications?
```

Sorry for jumping in here, but perhaps this suggestion is helpful (I didn't see the missing second _I_ at first either)

File: .rubocop.yml (hunk 14 above)
Comment (RachalCassity) [reply .rubocop.yml:67]: This pull request is still in draft. This was added due to a change in rubocop requirements but not be needed since rubocop was updated.

File: app/controllers/concerns/vet360/writeable.rb [hunk 15]
```diff
@@ -29,15 +34,15 @@ def invalidate_cache
       VAProfileRedis::Cache.invalidate(@current_user)
     end
 
-    private
-
     def build_record(type, params)
```
Comment (LindseySaari) [inline app/controllers/concerns/vet360/writeable.rb:37]: Should private be moved up to include `build_record` here? It seems it's only called from within the class and not externally

File: lib/va_profile/models/person.rb [hunk 16]
```diff
@@ -20,6 +20,18 @@ class Person < Base
       attribute :updated_at, Common::ISO8601Time
       attribute :vet360_id, String
 
+      CONTACT_INFO_CHANGE_TEMPLATE = Settings.vanotify.services.va_gov.template_id.contact_info_change
```
Comment (LindseySaari) [inline lib/va_profile/models/person.rb:23]: Can we put a new line between CONTACT_INFO_CHANGE_TEMPLATE and EMAIL_PERSONALISATIONS. It seems they run into each other

File: lib/va_profile/profile_information/service.rb [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+
+      def get_person
+        with_monitoring do
+          vet360_id_present!
+          raw_response = perform(:get, @user.vet360_id)
+          PersonResponse.from(raw_response)
+        end
+      rescue Common::Client::Errors::ClientError => e
+        if e.status == 404
+          log_exception_to_sentry(
```
Comment (LindseySaari) [inline lib/va_profile/profile_information/service.rb:42]: Let's not change this now, but I think we should try to avoid logging things directly to Sentry since this will likely be going out the door soon. Maybe for our final cleanup and refactor, we can take this on

**Review by jsimonVA — DISMISSED** [review]
Doesn't look like anything pertains to my team.

**Review by agravell047 — DISMISSED** [review]
No affect on MHV medications. LGTM

File: lib/va_profile/profile_information/service.rb (hunk 17 above)
Comment (RachalCassity) [reply lib/va_profile/profile_information/service.rb:42]: @LindseySaari I agree! We can add a rails logger instead :)

File: app/controllers/concerns/vet360/writeable.rb (hunk 15 above)
Comment (RachalCassity) [reply app/controllers/concerns/vet360/writeable.rb:37]: Moved back in private

File: lib/va_profile/models/person.rb (hunk 16 above)
Comment (RachalCassity) [reply lib/va_profile/models/person.rb:23]: Done!

File: spec/lib/va_profile/profile_information/service_spec.rb [hunk 18]
```diff
@@ -0,0 +1,858 @@
...
+
+  # describe '#get_person' do
+  # context 'when successful' do
+  #   it 'returns a status of 200' do
+  #     VCR.use_cassette('va_profile/profile_information/person_full', VCR::MATCH_EVERYTHING) do
+  #       response = subject.get_response
+  #       expect(response).to be_ok
+  #       expect(response.person).to be_a(VAProfile::Models::Person)
+  #     end
+  #   end
```
Comment (rmtolmach) [inline spec/lib/va_profile/profile_information/service_spec.rb:29]: I see a lot of tests in this file are commented out. That doesn't seem good 😰

File: lib/va_profile/profile_information/service.rb [hunk 19]
```diff
@@ -0,0 +1,127 @@
...
+      def submit(params)
+        config.submit(path(@user.edipi), params)
+      end
+
+      # Ensure http_verb is a symbol for the response request
+      def create_or_update_info(http_verb, record)
+        with_monitoring do
+          icn_with_aaid_present!
+
+          http_verb = http_verb.to_sym
```
Comment (rmtolmach) [inline lib/va_profile/profile_information/service.rb:73]: The comment at the top seems more appropriate on line 73. What do you think? We could also make it shorter: 
`# Convert to symbol for the response request`.
```suggestion
      def create_or_update_info(http_verb, record)
        with_monitoring do
          icn_with_aaid_present!

          http_verb = http_verb.to_sym # Ensure http_verb is a symbol for the response request
```

File: lib/va_profile/profile_information/transaction_response.rb [hunk 20]
```diff
@@ -0,0 +1,194 @@
...
+        return_val = super
+
+        log_error
+
+        return_val.response_body = @response_body
+        return_val
+      end
+
+      def changed_field
+        return :address unless response_body['tx_output']
```
Comment (LindseySaari) [inline lib/va_profile/profile_information/transaction_response.rb:51]: You could put `response_body['tx_output']` into a variable so it's not referenced multiple times

File: lib/va_profile/profile_information/transaction_response.rb [hunk 21]
```diff
@@ -0,0 +1,194 @@
...
+
+        address_pou = response_body['tx_output'][0]['address_pou']
```
Comment (LindseySaari) [inline lib/va_profile/profile_information/transaction_response.rb:53]: Should we check for ['address_pou'] being nil?

File: spec/lib/va_profile/profile_information/service_spec.rb (hunk 18 above)
Comment (LindseySaari) [reply spec/lib/va_profile/profile_information/service_spec.rb:29]: @RachalCassity Can you explain all the commented out tests

Comment (RachalCassity) [reply spec/lib/va_profile/profile_information/service_spec.rb:29]: @LindseySaari Those will be uncommented in the next PR.

File: lib/va_profile/profile_information/service.rb (hunk 19 above)
Comment (RachalCassity) [reply lib/va_profile/profile_information/service.rb:73]: I'll put the comment on 71 to reduce confusion

File: lib/va_profile/profile_information/transaction_response.rb (hunk 21 above)
Comment (RachalCassity) [reply lib/va_profile/profile_information/transaction_response.rb:53]: The case for address_pou does account if it is null.

**Comment by tpharrison** [comment]
@RachalCassity

Instead of:
- `VAProfile::ContactInformation::<class>`
- `VAProfile::ProfileInformation::<class>`

Would it be more clear to have something like?
- `VAProfile::ContactInformation::V1::<class>` (read/write)
- `VAProfile::ContactInformation::V2::<class>` (read/write)
- `VAProfile::ContactInformation::V3::<class>` (read/write)
- `VAProfile::Profile::V3::<class>` (read only)

File: lib/va_profile/models/permission.rb [hunk 22]
```diff
@@ -75,6 +75,22 @@ def self.build_from(body)
...
+        "#{user.vet360_id}/permissions/status/#{transaction_id}"
+      end
+
+      def self.send_change_notifications?
+        false
+      end
+
+      def contact_info_attr(*)
+        'text_permission'
+      end
```
Comment (rmtolmach) [inline lib/va_profile/models/permission.rb:93]: Why take `*` as an argument in this method if no arguments are being used?

File: lib/va_profile/models/email.rb [hunk 23]
```diff
@@ -64,6 +64,22 @@ def self.build_from(body)
           vet360_id: body['vet360_id']
         )
       end
+
+      def transaction_response_class
+        VAProfile::ProfileInformation::EmailTransactionResponse
+      end
+
+      def self.transaction_status_path(user, transaction_id)
+        "#{user.vet360_id}/emails/status/#{transaction_id}"
```
Comment (rmtolmach) [inline lib/va_profile/models/email.rb:73]: I see both `vet360_id` and `transaction_id` as attributes in this model. Is calling `user.` a necessary part of getting the `vet360_id`?

File: lib/va_profile/profile_information/configuration.rb [hunk 24]
```diff
@@ -0,0 +1,23 @@
...
+
+require 'va_profile/configuration'
+
+module VAProfile
+  module ProfileInformation
+    class Configuration < VAProfile::Configuration
+      self.read_timeout = VAProfile::Configuration::SETTINGS.profile_information.timeout || 30
+
+      def base_path
+        "#{VAProfile::Configuration::SETTINGS.url}/contact-information-hub/cuf/contact-information/v2"
```
Comment (rmtolmach) [inline lib/va_profile/profile_information/configuration.rb:11]: How did you know what to set this URL as?

File: spec/requests/va_profile/address_request_spec.rb [hunk 25]
```diff
@@ -11,6 +11,12 @@
   let(:headers_with_camel) { headers.merge('X-Key-Inflection' => 'camel') }
   let(:frozen_time) { Time.zone.local(2018, 6, 6, 15, 35, 55) }
 
+  if Flipper.enabled?(:va_profile_information_v3_service)
+    let(:cassette) { 'va_profile/profile_information/' }
+  else
+    let(:cassette) { 'va_profile/contact_information/' }
+  end
```
Comment (rmtolmach) [inline spec/requests/va_profile/address_request_spec.rb:18]: Nice 👍 this was a good idea.

File: spec/requests/va_profile/address_request_spec.rb [hunk 26]
```diff
@@ -24,12 +30,14 @@
     let(:address) { build(:va_profile_address, vet360_id: user.vet360_id) }
 
     it 'calls update_address' do
-      expect_any_instance_of(VAProfile::ContactInformation::Service).to receive(:update_address).and_call_original
-      VCR.use_cassette('va_profile/contact_information/put_address_success') do
-        post('/v0/profile/addresses/create_or_update', params: address.to_json, headers:)
+      # This can be removed after Contact Information is degraded
```
Comment (rmtolmach) [inline spec/requests/va_profile/address_request_spec.rb:33]:
```suggestion
      # This can be removed after Contact Information is deprecated
```

File: spec/requests/va_profile/email_address_request_spec.rb [hunk 27]
```diff
@@ -24,12 +30,14 @@
     let(:email) { build(:email, vet360_id: user.vet360_id) }
 
     it 'calls update_email' do
-      expect_any_instance_of(VAProfile::ContactInformation::Service).to receive(:update_email).and_call_original
-      VCR.use_cassette('va_profile/contact_information/put_email_success') do
-        post('/v0/profile/email_addresses/create_or_update', params: email.to_json, headers:)
+      # This can be removed after Contact Information is degraded
```
Comment (rmtolmach) [inline spec/requests/va_profile/email_address_request_spec.rb:33]:
```suggestion
      # This can be removed after Contact Information is deprecated
```

File: spec/requests/va_profile/permission_request_spec.rb [hunk 28]
```diff
@@ -23,13 +29,18 @@
   describe 'POST /v0/profile/permissions/create_or_update' do
     let(:permission) { build(:permission, vet360_id: user.vet360_id) }
 
-    it 'calls update_permission' do
-      expect_any_instance_of(VAProfile::ContactInformation::Service).to receive(:update_permission).and_call_original
-      VCR.use_cassette('va_profile/contact_information/put_permission_success') do
-        post('/v0/profile/permissions/create_or_update', params: permission.to_json, headers:)
+    context 'with a 200 response' do
+      it 'calls update_permission' do
+        # This can be removed after Contact Information is degraded
```
Comment (rmtolmach) [inline spec/requests/va_profile/permission_request_spec.rb:34]:
```suggestion
        # This can be removed after Contact Information is deprecated
```

File: spec/requests/va_profile/telephone_request_spec.rb [hunk 29]
```diff
@@ -23,12 +29,13 @@
     let(:telephone) { build(:telephone, vet360_id: user.vet360_id) }
 
     it 'calls update_telephone' do
-      expect_any_instance_of(VAProfile::ContactInformation::Service).to receive(:update_telephone).and_call_original
-      VCR.use_cassette('va_profile/contact_information/put_telephone_success') do
-        post('/v0/profile/telephones/create_or_update', params: telephone.to_json, headers:)
+      # This can be removed after Contact Information is degraded
```
Comment (rmtolmach) [inline spec/requests/va_profile/telephone_request_spec.rb:32]:
```suggestion
      # This can be removed after Contact Information is deprecated
```
